### DISCUSSION TITLE
Receiving Payments on Desktop Apps

### DISCUSSION THREAD

**Discussion by Serge-Mazur** [python-telegram-bot/python-telegram-bot#2804] [Q&A] [answered]
Hi.

I'm creating a bot to receive payments over Telegram and my searches led me to this example:
https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/paymentbot.py

I implemented it and it works with test payment system tokens, which is really nice. 🙂

My questions are:
1. Why does it work on mobile devices only?
According to this page "[Introducing Payments 2.0](https://core.telegram.org/bots/payments#introducing-payments-2-0)" bots `Accept payments from users on mobile or desktop apps.`
Is there a way in PTB to accept payments from any device too?

2. I tested [the official ShopBot](https://t.me/shopbot) and after my payment was processed I see a button "Receipt" with my checkout details.
You can see it as well here: https://i.imgur.com/aQbmQMP.png
Is it possible to do the same with the PTB library?

Thank you.
Serge

**Comment by Bibo-Joshi** [comment]
Hi. Glad to hear that the example for fine for you 👍

1. PTB doesn't know which device you use. That info is not provided by the Bot API. so if you can't make a payment through your desktop client, that sounds like an issue of that client to me.
2. [Here](https://core.telegram.org/bots/payments#8-checkout) it says 
    > If the invoice message was sent in the chat with @merchantbot, it becomes a Receipt in the UI for the user — they can open this receipt at any time and see all the details of the transaction:

    which to me sounds like that's a special feature of the example bots. I don't see any possibility in the [official api docs](https://core.telegram.org/bots/api) for such a button

**Reply by Serge-Mazur** [reply]
Thank you for your reply, @Bibo-Joshi.

Okay, now I see it too about my question 2. Then I considered it as closed.

Speaking of my question 1. I did some more testing with the `ShopBot` and the thing that allows me to pay from my desktop app is unticking `Webview`. I'm attaching a screenshot here:
https://i.imgur.com/5bTP2H6.png

However I didn't find how to modify the `webview` parameter in PTB. Is it possible? 🤔

**Reply by Bibo-Joshi** [reply]
the `Webview` thing apparently changes how the form for the credit card is displayed. Both versions work for me on TG desktop. But how the credit card form is displayed is the clients descision, I guess. wtih the bot api, you just send the invoice. The Bot API has no part in processing the payment method.

**Reply by Serge-Mazur** [reply]
That's odd. In my case when I tick `Webview` I can not finish my test payment on the desktop app. I see an error like on the screenshot here:
https://i.imgur.com/KxArLA3.png

And when I untick it then I'm able to finish my payment on the desktop app just fine.

So my tests gave me these results:
Mobile with Webview — Success
Mobile without Webview — Success
Desktop with Webview — Fail
Desktop without Webview — Success

Now I wonder how to switch the parameter `Webview` with Python.

**Reply by Bibo-Joshi** [reply]
> But how the credit card form is displayed is the clients descision, I guess. wtih the bot api, you just send the invoice. The Bot API has no part in processing the payment method.

Read as: "You can't.". I suggest to file a bug report for the desktop app that you're using.

**Reply by Serge-Mazur** [reply]
Oh, it's a bug apparently. So that means payments are supposed to work on any device and it could be fixed. I take it as good news. 🙂

I [searched](https://github.com/telegramdesktop/tdesktop/issues?q=webview) for the `Webview` issues and someone already created an issue about it 11 days ago:
https://github.com/telegramdesktop/tdesktop/issues/17290

How interesting.

**Comment by Serge-Mazur** [comment, marked as answer]
As of now in September 2024 the payments function without any errors. Tested on:
- Desktop Telegram (Win11)
- Mobile Telegram (Android 10)
- Mobile Telegram (iOS 17)

Used my test tokens from different payment providers, and also used my live token. Everything went fine.
It was, indeed, a bug. 🙂
